## Segmentation and Clustering of Neighborhood in Toronto, Canada.

### Introduction
    1.Web-scraping of the Toronto Neighborhood data using requests and BeautifulSoup Libraries.
    
    2.Performing Exploratory and Data analysis Using the Pandas and Numpy Library.
    
    3.Using Foursquare API Explore the Neighborhoods of Toronto location.
    
    4.Using Scikit-learn library to apply K-means Clustering alogrithm.  

## Table of Contents

1. Performing Web-scraping and Extracting Neighborhood and borough information from Wikipedia Page

2. Perfoming EDA 

3. Explore Neighborhoods in Toronto

4. Analyze Each Neighborhood

5. Cluster Neighborhoods

6. Examine Clusters    


## 1.  Performing Web-scraping and Extracting Neighborhood and borough information from Wikipedia Page

In [26]:
# Installing essential libraries in Web-scraping
# import the library we use to open the URL 
import lxml
import requests # library to handle requests
# Import the beautifulSoup library so we can parse HTML and XML documents
from bs4 import BeautifulSoup as BS

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modulesl
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [58]:
# Storing the URL of wikipedia wiki_url
wiki_url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# Getting the source code of the HTML page from the URL using the request library
source = requests.get(wiki_url).text

# Converting the HTML souce code to the beautiful soup format
soup = BS(source,'html.parser')

# Reading the table Information and storing in the database 
df_old = pd.read_html(wiki_url)

# Using the find method to find the exact borough information 
df_old[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

In [60]:
# Checking the  database for all the Unique Borough names and total number of neighborhood
df = df_old[0].copy()
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

## 2. Perfoming EDA 
Neighborhood has a total of 3 boroughs and 177 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 3 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood.

### Data Handling and Cleaning

In [61]:
# Checking the dataFrame for missing Values
df.isna().sum()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

In [66]:
# dropping the rows having 'Not assigned' in Borough column
df = df[~(df['Borough'] == 'Not assigned')]
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [80]:
# Cheking again for any missing or null values
df.isna().sum()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

In [81]:
df.shape

(103, 3)

In [10]:
# Extracting Neihborhood names having values in 'name/name' format and 'name-name' format
def extract(x):
    x = x.split('–')[0]
    x = x.split('/')[0]
    return x
df.Name = df.Name.apply(extract)

### Using geopy library to get the latitudes and longitudes of each Neighborhood

43.695403
-79.293099


In [43]:
# Creating a new DataFrame to store the latitudes and longitudes of the each neighborhood
column_names = [  'Borough', 'Neighborhood', 'Latitude', 'Longitude'] 
Neighborhood = pd.DataFrame(columns=column_names)
Neighborhood

# We will define a instance of the geopy library as an tor_agent
# We cannot use the Nominatim method for large number of values hence we divide the dataset into two parts

# First Part of the dataset
for name , bor in zip(df.Name[0:70], df.Borough[0:70]):  
    try:
        address = f'{name} , Toronto'
        geolocator = Nominatim(user_agent="tor_agent")
        location = geolocator.geocode(address)
        Neighborhood = Neighborhood.append({
                                            'Borough':bor,
                                            'Neighborhood':name,
                                            'Latitude':location.latitude,
                                            'Longitude':location.longitude
                                           },ignore_index=True)
    except:
        Neighborhood = Neighborhood.append({
                                            'Borough':bor,
                                            'Neighborhood':name,
                                            'Latitude':np.NaN,
                                            'Longitude':np.NaN
                                           },ignore_index=True)

# Second Part of the dataset
for name , bor in zip(df.Name[70:], df.Borough[70:]):  
    try:
        address = f'{name}, Toronto'
        geolocator = Nominatim(user_agent="tor_agent")
        location = geolocator.geocode(address)
        Neighborhood = Neighborhood.append({
                                            'Borough':bor,
                                            'Neighborhood':name,
                                            'Latitude':location.latitude,
                                            'Longitude':location.longitude
                                           },ignore_index=True)
    except:
        Neighborhood = Neighborhood.append({
                                            'Borough':bor,
                                            'Neighborhood':name,
                                            'Latitude':np.NaN,
                                            'Longitude':np.NaN
                                           },ignore_index=True)

# Cheking for any null values        
Neighborhood.isna().sum()

Borough         0
Neighborhood    0
Latitude        6
Longitude       6
dtype: int64

In [44]:
# Removing the names not having Proper coordinates
Neighborhood = Neighborhood[~Neighborhood.Latitude.isna()]

# Checking for any Missing values or Null values
Neighborhood.isna().sum()

Borough         0
Neighborhood    0
Latitude        0
Longitude       0
dtype: int64

In [32]:
# Conforming after map visualization given below latitude location in not proper for Niagara Neighborhood
Neighborhood = Neighborhood[~ (Neighborhood['Neighborhood'] == 'Niagara')]
Neighborhood.shape

(169, 4)

In [45]:
Neighborhood.head()

Borough       Neighborhood   Latitude  Longitude
0  East York      Crescent Town  43.695403 -79.293099
1  East York  Governor's Bridge  43.689423 -79.369426
2  East York            Leaside  43.704798 -79.368090
3  East York           O'Connor  43.750275 -79.317901
4  East York      Old East York  43.670862 -79.372792

The Final Extracted dataset has 3 boroughs and 169 neighborhoods.

In [46]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(Neighborhood['Borough'].unique()),
        Neighborhood.shape[0]
    )
)

The dataframe has 3 boroughs and 167 neighborhoods.


## 3. Explore Neighborhoods in Toronto

In [36]:
# Using geopy.geocoders librarys Nominatim method to convert address to latitudes and longitudes
neighborhood_address = "toronto" # 
geolocator = Nominatim(user_agent="tor_agent")
location = geolocator.geocode(neighborhood_address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

43.6534817 -79.3839347


In [47]:
# Creating a map of Toronto Neighborhood using Folium Library and latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, long, borough, neigh in zip(Neighborhood['Latitude'], 
                                     Neighborhood['Longitude'], 
                                     Neighborhood['Borough'], 
                                     Neighborhood['Neighborhood']):
    label = '{}, {}'.format(neigh, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=label,
        colors='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.4,
        parse_html=False).add_to(map_toronto)
map_toronto

In [ ]:
## For generating N values of random color in 
# x_coordinates = [0, 1, 2, 3, 4, 5]
# y_coordinates = [0, 1, 2, 3, 4, 5]

# for x, y in zip(x_coordinates, y_coordinates):
#     rgb = (random.random(), random.random(), random.random())
#     plt.scatter(x, y, c=[rgb])

### Exploring the dataset for the Scarborough borough only

In [50]:
Scarborough_Neighborhood = Neighborhood[Neighborhood['Borough'] == 'Scarborough']
Scarborough_Neighborhood.reset_index(drop=True)

Borough             Neighborhood   Latitude  Longitude
0    Scarborough                Agincourt  43.785353 -79.278549
1    Scarborough           Alexandra Park  43.650787 -79.404318
2    Scarborough                  Allenby  43.711351 -79.553424
3    Scarborough                 Amesbury  43.706162 -79.483492
4    Scarborough           Armour Heights  43.743944 -79.430851
5    Scarborough                  Banbury  43.742796 -79.369957
6    Scarborough           Bathurst Manor  43.665519 -79.411937
7    Scarborough      Bay Street Corridor  43.667342 -79.388457
8    Scarborough          Bayview Village  43.769197 -79.376662
9    Scarborough           Bayview Woods   43.798127 -79.382973
10   Scarborough             Bedford Park  43.737388 -79.410925
11   Scarborough                  Bendale  43.753520 -79.255336
12   Scarborough              Birch Cliff  43.691805 -79.264494
13   Scarborough       Bloor West Village  43.649318 -79.484436
14   Scarborough          Bracondale Hill  43.676195 -79.428016
15   Scarborough                  Branson  43.771426 -79.447287
16   Scarborough              Bridle Path  43.738151 -79.372511
17   Scarborough                 Brockton  43.650917 -79.440022
18   Scarborough              Cabbagetown  43.664473 -79.366986
19   Scarborough             Caribou Park  43.720850 -79.415274
20   Scarborough         Carleton Village  43.670701 -79.453299
21   Scarborough                Casa Loma  43.678101 -79.409416
22   Scarborough          Chaplin Estates  43.702598 -79.403270
23   Scarborough            Christie Pits  43.667139 -79.422766
24   Scarborough     Church and Wellesley  43.665524 -79.383801
25   Scarborough                 Clairlea  43.708823 -79.295986
26   Scarborough               Cliffcrest  43.721836 -79.236214
27   Scarborough                Cliffside  43.711170 -79.248177
28   Scarborough                 Corktown  43.657370 -79.356513
29   Scarborough             Cricket Club  43.830638 -79.224609
30   Scarborough                Davenport  43.671545 -79.448322
31   Scarborough               Davisville  43.697936 -79.397291
32   Scarborough                Deer Park  43.688090 -79.394093
33   Scarborough       Discovery District  43.657556 -79.389480
34   Scarborough      Distillery District  43.650295 -79.359540
35   Scarborough                Don Mills  43.775347 -79.345944
36   Scarborough       Don Valley Village  43.792673 -79.354722
37   Scarborough              Dorset Park  43.752847 -79.282067
38   Scarborough          Dovercourt Park  43.665223 -79.433537
39   Scarborough                Downsview  43.749299 -79.462248
40   Scarborough           Dufferin Grove  43.653632 -79.426439
41   Scarborough               Earlscourt  43.676277 -79.450572
42   Scarborough            East Danforth  43.686433 -79.300355
43   Scarborough            Eglinton East  43.739465 -79.232100
44   Scarborough    Elia (Jane and Finch)  43.757253 -79.517697
45   Scarborough         Fashion District  43.645456 -79.394994
46   Scarborough       Financial District  43.648664 -79.381540
47   Scarborough          Flemingdon Park  43.718432 -79.333204
48   Scarborough              Forest Hill  43.693559 -79.413902
49   Scarborough                Fort York  43.638992 -79.403445
50   Scarborough          Garden District  43.656500 -79.377114
51   Scarborough                Glen Park  43.709002 -79.443798
52   Scarborough              Grange Park  43.652197 -79.392319
53   Scarborough             Graydon Hall  43.762863 -79.342667
54   Scarborough                Guildwood  43.755225 -79.198229
55   Scarborough          Harbord Village  43.661522 -79.409745
56   Scarborough            Harbourfront   43.640080 -79.380150
57   Scarborough                  Harwood  43.677044 -79.476151
58   Scarborough               Henry Farm  43.769509 -79.354296
59   Scarborough          High Park North  43.657383 -79.470961
60   Scarborough           Highland Creek  43.790117 -79.173334
61   Scarborough                Hillcrest

## Exploring the Scarborough of Borough
#### As we did for the all the neighborhood lets Visualize the Neiborhood of Scarborough


In [55]:
# Using geopy.geocoders librarys Nominatim method to convert address to latitudes and longitudes
neighborhood_address = "Scarborough, Toronto" # 
geolocator = Nominatim(user_agent="scarborough_agent")
location = geolocator.geocode(neighborhood_address)
latitude = location.latitude
longitude = location.longitude
print('The latitude and longitude of Scarborough,Toronto in  {}, {}'.format(latitude, longitude) )

The latitude and longitude of Scarborough,Toronto in  43.773077, -79.257774


In [57]:
# Creating a map of Toronto Neighborhood using Folium Library and latitude and longitude values
map_scarborough = folium.Map(location=[latitude, longitude], zoom_start=10)

for lat, long, borough, neigh in zip(Scarborough_Neighborhood['Latitude'], 
                                     Scarborough_Neighborhood['Longitude'], 
                                     Scarborough_Neighborhood['Borough'], 
                                     Scarborough_Neighborhood['Neighborhood']):
    label = '{}, {}'.format(neigh, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=label,
        colors='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.4,
        parse_html=False).add_to(map_scarborough)
map_scarborough